pkg/csi/service/vanilla/controller_topology_test.go

```diff
@@ -0,0 +1,501 @@
+			VcenterManager: vcManager,
+		}
+
+		var k8sClient clientset.Interface
+		if k8senv := os.Getenv("KUBECONFIG"); k8senv != "" {
+			k8sClient, err = k8s.CreateKubernetesClientFromConfig(k8senv)
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			k8sClient = testclient.NewSimpleClientset()
+		}
+
+		nodeManager := &FakeNodeManagerTopology{
+			k8sClient: k8sClient}
+		err = nodeManager.Initialize(ctxtopology)
+		if err != nil {
+			t.Fatalf("Failed to initialize the node manager, err= =%v", err)
+		}
+
+		mockNodeLabels, err := generateNodeLabels(ctxtopology, vcenter)
+		if err != nil {
+			t.Fatalf("Failed to add mock node labels, err = %v", err)
+		}
+
+		c := &controller{
+			manager: manager,
+			nodeMgr: nodeManager,
+			authMgr: &FakeAuthManager{
+				vcenter: vcenter,
+			},
+			topologyMgr: &FakeTopologyManager{
+				nodeLabels: mockNodeLabels,
+			},
+		}
+		commonco.ContainerOrchestratorUtility, err =
+			unittestcommon.GetFakeContainerOrchestratorInterface(common.Kubernetes)
+		if err != nil {
+			t.Fatalf("Failed to create co agnostic interface. err=%v", err)
+		}
+		controllerTestInstanceTopology = &controllerTestTopology{
+			controller:     c,
+			config:         config,
+			vcenter:        vcenter,
+			operationStore: fakeOpStore,
+		}
+	})
+	return controllerTestInstanceTopology
+}
+
+// Test CreateVolume functionality with specific AccessibilityRequirements
+func TestCreateVolumeWithAccessibilityRequirements(t *testing.T) {
+	// Create context.
+	ct := getControllerTestWithTopology(t)
+
+	// Create volume
+	params := make(map[string]string)
+	capabilities := []*csi.VolumeCapability{
+		{
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+			},
+		},
+	}
+
+	reqCreate := &csi.CreateVolumeRequest{
+		Name: testVolumeName + "-" + uuid.New().String(),
+		CapacityRange: &csi.CapacityRange{
+			RequiredBytes: 1 * common.GbInBytes,
+		},
+		Parameters:         params,
+		VolumeCapabilities: capabilities,
+		AccessibilityRequirements: &csi.TopologyRequirement{
+			Requisite: []*csi.Topology{
+				{
+					Segments: map[string]string{
+						"topology.csi.vmware.com/k8s-zone": "zone-1",
+					},
+				},
+			},
+			Preferred: []*csi.Topology{},
+		},
+	}
+
+	respCreate, err := ct.controller.CreateVolume(ctxtopology, reqCreate)
+	if err != nil {
+		t.Fatal(err)
+	}
+	volID := respCreate.Volume.VolumeId
+
+	queryFilter := cnstypes.CnsQueryFilter{
+		VolumeIds: []cnstypes.CnsVolumeId{
+			{
+				Id: volID,
+			},
+		},
+	}
+	queryResult, err := ct.vcenter.CnsClient.QueryVolume(ctxtopology, queryFilter)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(queryResult.Volumes) != 1 && queryResult.Volumes[0].VolumeId.Id != volID {
+		t.Fatalf("failed to find the newly created volume with ID: %s", volID)
+	}
+
+	// Delete volume
+	reqDelete := &csi.DeleteVolumeRequest{
+		VolumeId: volID,
+	}
+	_, err = ct.controller.DeleteVolume(ctxtopology, reqDelete)
```

Comment:
Can we split this test into two. Create and Delete.

One unit test for `Test_CreateVolume_WithAccessibilityRequirements` and another unit test for `Test_DeleteVolume`

Reply:
We need to anyway clean the volume created and we are calling both these functions together in all other unit tests, so keeping this as it is for consistency.